Scio pipelines should be executable with a classpath referencing a local repository

My team has upgraded from scio 9.3 to scio 10.4 and now we are unable to execute pipelines straight from intellij because our entire `~/.m2` cache is being filtered out in local execution. In the intellij runtime, most classpath jars are in `user.home`. When we try to run pipelines straight from IJ, we get `NoClassDefFoundError` because the entire classpath is filtered out of `PipelineOptions.filesToStage`.

Example exception in a dataflow job:
```
Caused by: java.lang.NoClassDefFoundError: org/apache/beam/sdk/options/PipelineOptions
```

Relevant logs in the pipeline launch process:
```
{"timestamp":"2021-09-21T20:25:51.510Z","level":"INFO","thread":"main","logger":"org.apache.beam.runners.dataflow.util.PackageUtil","message":"Uploading 4 files from PipelineOptions.filesToStage to staging location to prepare for execution.","context":"default"}
```

^ As you can see, only 4 files are uploaded (our classpath should contain ~200 dependencies)

We've tracked this down to the difference in behavior between 9.3 and 10.4. This is the functionality in 9.3
https://github.com/spotify/scio/blob/v0.9.3/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala#L141

```
CoreSysProps.UserDir.value
> /Users/zacp/path/to/repo/root
```

And this is in 10.4
https://github.com/spotify/scio/blob/main/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala#L150

```
sys.props("user.home")
> /Users/zacp
```

So, anything in my `~` directory is not being staged. This includes the entirety of `~/.m2` and `~/.ivy`. We're manually setting our `filesToStage` for now, but it seems like this is possibly undesired default behavior.

--- a/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
@@ -142,12 +142,16 @@ private object RunnerContext {
     classLoader: ClassLoader
   ): Iterable[String] = {
     val sanitize: String => String = _.replace("\\", "/")
-    val matchesEnvDir: String => Boolean =
-      _.matches(s"${sanitize(sys.props("user.home"))}/\\..+/.+")
-    val classPathJars = PipelineResources
+    val sanitizedHome = sanitize(sys.props("user.home"))
+    val matchesWhitelistDir: String => Boolean = s => s.matches(s"${sanitizedHome}/\\.(ivy2|m2)/.+") || s.matches(s"${sanitizedHome}/\\.cache/coursier/.+")
+    val matchesEnvDir: String => Boolean = _.matches(s"${sanitizedHome}/\\..+/.+")
+    val classPathJars = org.apache.beam.runners.core.construction.resources.PipelineResources
       .detectClassPathResourcesToStage(classLoader, pipelineOptions)
       .asScala
-      .filterNot(sanitize.andThen(matchesEnvDir))
+      .filterNot { s =>
+        val x = sanitize(s)
+        matchesEnvDir(x) && !matchesWhitelistDir(x)
+      }
 
     logger.debug(s"Classpath jars: ${classPathJars.mkString(":")}")
 

--- a/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
@@ -29,6 +29,7 @@ import com.spotify.scio.options.ScioOptions
 import com.spotify.scio.testing._
 import com.spotify.scio.util._
 import com.spotify.scio.values._
+import org.apache.beam.runners.core.construction.resources.PipelineResources
 import org.apache.beam.sdk.Pipeline.PipelineExecutionException
 import org.apache.beam.sdk.PipelineResult.State
 import org.apache.beam.sdk.extensions.gcp.options.GcsOptions
@@ -149,7 +150,7 @@ private object RunnerContext {
     pipelineOptions: PipelineOptions,
     classLoader: ClassLoader
   ): Iterable[String] = {
-    val classPathJars = org.apache.beam.runners.core.construction.resources.PipelineResources
+    val classPathJars = PipelineResources
       .detectClassPathResourcesToStage(classLoader, pipelineOptions)
       .asScala
       .filterNot(sanitizePath.andThen(isNonRepositoryEnvDir))

--- a/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
@@ -143,7 +143,10 @@ private object RunnerContext {
   ): Iterable[String] = {
     val sanitize: String => String = _.replace("\\", "/")
     val sanitizedHome = sanitize(sys.props("user.home"))
-    val matchesWhitelistDir: String => Boolean = s => s.matches(s"${sanitizedHome}/\\.(ivy2|m2)/.+") || s.matches(s"${sanitizedHome}/\\.cache/coursier/.+")
+    val matchesWhitelistDir: String => Boolean = s =>
+      s.matches(s"${sanitizedHome}/\\.(ivy2|m2)/.+") || s.matches(
+        s"${sanitizedHome}/\\.cache/coursier/.+"
+      )
     val matchesEnvDir: String => Boolean = _.matches(s"${sanitizedHome}/\\..+/.+")
     val classPathJars = org.apache.beam.runners.core.construction.resources.PipelineResources
       .detectClassPathResourcesToStage(classLoader, pipelineOptions)

--- a/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
@@ -135,26 +135,24 @@ private object RunnerContext {
     finalLocalArtifacts
   }
 
+  private val sanitizePath: String => String = _.replace("\\", "/")
+
+  private[scio] def isNonRepositoryEnvDir(s: String): Boolean = {
+    val sanitizedUserHome: String = sanitizePath(sys.props("user.home"))
+    s.matches(s"${sanitizedUserHome}/\\..+/.+") && !s.matches(
+      s"${sanitizedUserHome}/\\.(ivy2|m2|cache/coursier)/.+"
+    )
+  }
+
   /** Borrowed from DataflowRunner. */
   private[this] def detectClassPathResourcesToStage(
     pipelineOptions: PipelineOptions,
     classLoader: ClassLoader
   ): Iterable[String] = {
-    val sanitize: String => String = _.replace("\\", "/")
-    val sanitizedHome = sanitize(sys.props("user.home"))
-    val matchesWhitelistDir: String => Boolean = s =>
-      s.matches(s"${sanitizedHome}/\\.(ivy2|m2)/.+") || s.matches(
-        s"${sanitizedHome}/\\.cache/coursier/.+"
-      )
-    val matchesEnvDir: String => Boolean = _.matches(s"${sanitizedHome}/\\..+/.+")
     val classPathJars = org.apache.beam.runners.core.construction.resources.PipelineResources
       .detectClassPathResourcesToStage(classLoader, pipelineOptions)
       .asScala
-      .filterNot { s =>
-        val x = sanitize(s)
-        matchesEnvDir(x) && !matchesWhitelistDir(x)
-      }
-
+      .filterNot(sanitizePath.andThen(isNonRepositoryEnvDir))
     logger.debug(s"Classpath jars: ${classPathJars.mkString(":")}")
 
     classPathJars

--- a/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
+++ b/scio-core/src/main/scala/com/spotify/scio/ScioContext.scala
@@ -29,7 +29,6 @@ import com.spotify.scio.options.ScioOptions
 import com.spotify.scio.testing._
 import com.spotify.scio.util._
 import com.spotify.scio.values._
-import org.apache.beam.runners.core.construction.resources.PipelineResources
 import org.apache.beam.sdk.Pipeline.PipelineExecutionException
 import org.apache.beam.sdk.PipelineResult.State
 import org.apache.beam.sdk.extensions.gcp.options.GcsOptions

--- a/scio-test/src/test/scala/com/spotify/scio/ScioContextTest.scala
+++ b/scio-test/src/test/scala/com/spotify/scio/ScioContextTest.scala
@@ -261,4 +261,17 @@ class ScioContextTest extends PipelineSpec {
     noException shouldBe thrownBy(sc.run())
   }
 
+  "RunnerContext" should "include ~/.m2, ~/.ivy2, and ~/.cache/coursier dirs, but not other env dirs" in {
+    val userDir = sys.props("user.home").replace("\\", "/")
+    val paths1 = List(
+      s"$userDir/.m2/repository/com/spotify/foo/0.0.1/foo-0.0.1.jar",
+      s"$userDir/.ivy2/local/com/spotify/foo/0.0.1/jars/foo-0.0.1.jar",
+      s"$userDir/.cache/coursier/v1/https/repo1.maven.org/maven2/com/spotify/foo/0.0.1/foo-0.0.1.jar"
+    )
+    val paths2 = paths1 :+ s"$userDir/.env/a/b/c/foo-0.0.1.jar"
+
+    paths1.filterNot(RunnerContext.isNonRepositoryEnvDir) should contain theSameElementsAs paths1
+    paths2.filterNot(RunnerContext.isNonRepositoryEnvDir) should contain theSameElementsAs paths1
+  }
+
 }

